### DISCUSSION TITLE
ValueError: <class 'pydantic.networks.EmailStr'> has no matching SQLAlchemy type

### DISCUSSION THREAD

**Discussion by ahaly-compass** [fastapi/sqlmodel#730] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from pydantic import EmailStr
from sqlmodel import Field, SQLModel


class UserBase(SQLModel):
    name: str = Field(index=True)
    first_name: str
    last_name: str
    email: EmailStr  = Field(unique=True, index=True)
    active: bool = Field(default=True)


# # Database model, database table inferred from class name
class User(UserBase, table=True):
    __tablename__ = "users"
    id: int | None = Field(default=None, primary_key=True)
    hashed_password: str
    # tickets: list["Ticket"] = Relationship(back_populates="ticket")
```

### Description

* Starting the app I get the error

ValueError: <class 'pydantic.networks.EmailStr'> has no matching SQLAlchemy type

relevant dependencies:
* fastapi = "^0.104.1"
* uvicorn = "^0.24.0.post1"
* sqlmodel = "^0.0.14"
* psycopg = {extras = ["binary"], version = "^3.1.14"}

### Operating System

Linux

### Operating System Details

Ubuntu 22.04 WSL

### SQLModel Version

0.0.14

### Python Version

3.12

**Comment by rmasters** [comment, marked as answer]
I've run into this a few times. The fix is to pass the SQLAlchemy data type directly - this is `sqlalchemy.String()`, but in this case I'm using sqlmodel's own `AutoString` wrapper type:

```diff
-from sqlmodel import Field, SQLModel
+from sqlmodel import Field, SQLModel, AutoString

class UserBase(SQLModel):
    name: str = Field(index=True)
    first_name: str
    last_name: str
-   email: EmailStr  = Field(unique=True, index=True)
+   email: EmailStr  = Field(unique=True, index=True, sa_type=AutoString)
    active: bool = Field(default=True)
```

The root cause of this is `EmailStr` does not subclass a type that SQLModel knows how to convert to a SQLAlchemy type (i.e. it's not `class EmailStr(str)`. For example, Pydantic's [`CountryAlpha2`][countryalpha2] _does_ subclass `str`, so it works out of the box. 

[`sqlmodel.get_sqlalchemy_type()`][get_sqlalchemy_type] handles conversion from primitives like int and str to Float and AutoString, as well as helpers for a few common types like IP addresses. I've proposed adding `EmailStr` to this list of helpers in #762.

[get_sqlalchemy_type]: https://github.com/tiangolo/sqlmodel/blob/fe497adf0c0793c9a88cf2cfef125b6bdd498851/sqlmodel/main.py#L554
[countryalpha2]: https://github.com/pydantic/pydantic-extra-types/blob/0ce9f207a1e09a862287ab77512f0060c1625223/pydantic_extra_types/country.py#L70

**Reply by ahaly-compass** [reply]
I also found this solution:
`email: EmailStr = Field(sa_column=Column(String, index=True, unique=True))`

But yours is the SQLModel solution.
Thanks.

**Reply by salahelfarissi** [reply]
Thank you for your answer!

**Reply by ChrisNi888** [reply]
Yes, you are right. It works for HttpUrl:
`avatar: Optional[HttpUrl] = Field(sa_type=AutoString)`
Thanks!

**Reply by vinodkumars** [reply]
I tried this approach for HttpUrl but I get a SqlAlchemy error on runtime
```
sqlalchemy.exc.ProgrammingError: (psycopg2.ProgrammingError) can't adapt type 'pydantic_core._pydantic_core.Url'
[SQL: INSERT INTO "user" (id, created_at, updated_at, deleted_at, full_name, email, is_email_verified, image_url) VALUES (%(id)s, %(created_at)s, %(updated_at)s, %(deleted_at)s, %(full_name)s, %(email)s, %(is_email_verified)s, %(image_url)s)]
[parameters: {'id': '6hYri22AfrBViz1QY0gDY', 'created_at': datetime.datetime(2024, 5, 29, 22, 43, 35, 119940, tzinfo=datetime.timezone.utc), 'updated_at': datetime.datetime(2024, 5, 29, 22, 43, 35, 119947, tzinfo=datetime.timezone.utc), 'deleted_at': None, 'full_name': 'string', 'email': 'vinod@canihelp.ai', 'is_email_verified': False, 'image_url': Url('https://example.com/')}]
```

**Reply by mike-500** [reply]
If it's still relevant, for `HttpUrl` I use a custom `TypeDecorator` from `sqlalchemy.types`.

Full example:
```python
from datetime import datetime
from typing import Optional

from pydantic import ConfigDict, HttpUrl
from pydantic.alias_generators import to_camel
from sqlalchemy import func
from sqlalchemy.types import String, TypeDecorator, DateTime
from sqlmodel import Field, SQLModel

from .config import settings


class HttpUrlType(TypeDecorator):
    impl = String(2083)
    cache_ok = True
    python_type = HttpUrl

    def process_bind_param(self, value, dialect) -> str:
        return str(value)

    def process_result_value(self, value, dialect) -> HttpUrl:
        return HttpUrl(url=value)

    def process_literal_param(self, value, dialect) -> str:
        return str(value)


class UrlBase(SQLModel):
    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
        from_attributes=True,
    )

    url: HttpUrl = Field(
        index=True,
        unique=True,
        nullable=False,
        sa_type=HttpUrlType,
    )


class Url(UrlBase, table=True):
    __tablename__ = "urls"

    id: str = Field(
        max_length=settings.url_id_length,
        min_length=settings.url_id_length,
        primary_key=True,
        nullable=False,
    )
    visits: int = Field(default=0)
    created_at: Optional[datetime] = Field(
        default=None,
        sa_type=DateTime(timezone=True),
        sa_column_kwargs={
            "server_default": func.now(),  # pylint: disable=E1102
        },
    )
    updated_at: Optional[datetime] = Field(
        default=None,
        sa_type=DateTime(timezone=True),
        sa_column_kwargs={
            "onupdate": func.now(),  # pylint: disable=E1102
        },
    )


class UrlCreate(UrlBase):
    pass


class UrlRead(UrlBase):
    id: str
    visits: int
    created_at: datetime
    updated_at: Optional[datetime]

```
